Add trait for serialization

kampersanda opened the pull request:
This PR added traits for serializing/deserializing data structures to simplify codes.

vbkaisetsu commented on src/bytewise.rs:
```diff
@@ -11,6 +11,9 @@ use alloc::vec::Vec;
 use crate::errors::{DaachorseError, Result};
 use crate::intpack::{U24nU8, U24};
+use crate::serializer::{
+    deserialize_vec, serialize_slice, serialized_bytes, Deserialize, Serialize,
```
Could you follow the idiomatic way to bring fuctions?
https://doc.rust-lang.org/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html#creating-idiomatic-use-paths
> Specifying the parent module when calling the function makes it clear that the function isn’t locally defined while still minimizing repetition of the full path.

vbkaisetsu reviewed (commented):
I think the API should be as follows.

* Merge `Serialize` and `Deserialize` into the `Serializable` trait.
* Rename `to_vec` to `serialize_to_vec` because such name is already used in the slice and could be confusing.
* Remove the glogal `serialize_slice()` function, and implement `Serializable` in `Vec<S>`.
* Implement `serialized_bytes()` as the function of `Serializable`.

kampersanda commented:
@vbkaisetsu Thank you for your suggestions. I've modified codes following them.
I also implemented `Serializable` for `MatchKind` and `CodeMapper` for consistency.
Could you review them again?

vbkaisetsu commented on src/serializer.rs:
```diff
@@ -0,0 +1,72 @@
+        (dst, src)
+    }
+
+    fn serialized_bytes(&self) -> usize {
+        mem::size_of::<u32>() + mem::size_of::<S>() * self.len()
+    }
```
This is incorrect
```rust
let v1 = vec![0u32; 10];
dbg!(v1.serialized_bytes());
let v2 = vec![v1; 10];
dbg!(v2.serialized_bytes());
```
```
v1.serialized_bytes() = 44
v2.serialized_bytes() = 244
```

vbkaisetsu commented on src/serializer.rs:
```diff
@@ -0,0 +1,72 @@
+
+    fn deserialize_from_slice(src: &[u8]) -> (Self, &[u8]);
+
+    fn serialized_bytes(&self) -> usize {
+        mem::size_of::<Self>()
+    }
```
`size_of()` depends on its layout.

kampersanda commented:
@vbkaisetsu I fixed as follows:

- Specialized `Serializable` for `Vec<S>` as `SerializableVec`.
- Added tests
- Implemented `Serializable::serialized_bytes` in each specialization without shared behavior.

vbkaisetsu commented on src/lib.rs:
```diff
@@ -343,3 +358,41 @@ impl From<MatchKind> for u8 {
+
+    #[inline(always)]
+    fn serialized_bytes() -> usize {
+        core::mem::size_of::<Self>()
```
ditto

vbkaisetsu commented on src/bytewise.rs:
```diff
@@ -784,23 +754,34 @@ impl State {
+
+    #[inline(always)]
+    fn serialized_bytes() -> usize {
+        mem::size_of::<Self>()
```
ditto

kampersanda commented:
@vbkaisetsu To keep consistency to the implementations of `serialize_to_vec` and `deserialize_from_slice`, I fixed `serialized_bytes` to use `Serializable::serialized_bytes()`.

vbkaisetsu commented on src/bytewise.rs:
```diff
@@ -994,4 +977,29 @@ mod tests {
+        let mut opos_ch = U24nU8::default();
+        opos_ch.set_a(U24::try_from(57).unwrap());
+        opos_ch.set_b(77);
+        crate::serializer::tests::test_common(State {
```
Could you not hide the test code inside the function?

In unit testing, it is often said DAMP not DRY.
It is not clear what the behavior of `test_common` is, so we need to move on to the definition of this function.
https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests

kampersanda replied:
I removed `test_common` https://github.com/daac-tools/daachorse/pull/68/commits/32618fa25af522d9bb9aaece0f6923a135e0352f
I learned a lot, thanks!

vbkaisetsu commented on src/serializer.rs:
```diff
@@ -0,0 +1,125 @@
+
+    #[test]
+    fn test_u32() {
+        let x = 42u32;
+        let mut data = vec![];
+        x.serialize_to_vec(&mut data);
+        assert_eq!(data.len(), u32::serialized_bytes());
+        let (y, rest) = u32::deserialize_from_slice(&data);
+        assert!(rest.is_empty());
```
```suggestion
        let x = 0x01234567u32;
        let mut data = vec![];
        x.serialize_to_vec(&mut data);
        assert_eq!(vec![0x67, 0x45, 0x23, 0x01], data);
        assert_eq!(4, u32::serialized_bytes());
        data.push(42);
        let (y, rest) = u32::deserialize_from_slice(&data);
        assert_eq!(&[42], rest);
```

vbkaisetsu commented on src/serializer.rs:
```diff
@@ -0,0 +1,125 @@
+
+    #[test]
+    fn test_nzu32() {
+        let x = NonZeroU32::new(42u32);
+        let mut data = vec![];
+        x.serialize_to_vec(&mut data);
+        assert_eq!(data.len(), Option::<NonZeroU32>::serialized_bytes());
+        let (y, rest) = Option::<NonZeroU32>::deserialize_from_slice(&data);
+        assert!(rest.is_empty());
```
```suggestion
        let x = NonZeroU32::new(0x01234567u32);
        let mut data = vec![];
        x.serialize_to_vec(&mut data);
        assert_eq!(vec![0x67, 0x45, 0x23, 0x01], data);
        assert_eq!(4, Option::<NonZeroU32>::serialized_bytes());
        data.push(42);
        let (y, rest) = Option::<NonZeroU32>::deserialize_from_slice(&data);
        assert_eq!(&[42], rest);
```

vbkaisetsu commented on src/serializer.rs:
```diff
@@ -0,0 +1,125 @@
+
+    #[test]
+    fn test_vec_u32() {
+        let x = vec![42u32; 10];
+        let mut data = vec![];
+        x.serialize_to_vec(&mut data);
+        assert_eq!(data.len(), x.serialized_bytes());
+        let (y, rest) = Vec::<u32>::deserialize_from_slice(&data);
+        assert!(rest.is_empty());
```
```suggestion
        let x = vec![0x01234567u32, 0x89abcdefu32, 0x02468aceu32];
        let mut data = vec![];
        x.serialize_to_vec(&mut data);
        assert_eq!(
            vec![
                0x03, 0x00, 0x00, 0x00, // len
                0x67, 0x45, 0x23, 0x01, // item 1
                0xef, 0xcd, 0xab, 0x89, // item 2
                0xce, 0x8a, 0x46, 0x02, // item 3
            ],
            data
        );
        assert_eq!(16, x.serialized_bytes());
        data.push(42);
        let (y, rest) = Vec::<u32>::deserialize_from_slice(&data);
        assert_eq!(&[42], rest);
```

vbkaisetsu reviewed (commented):
I suggest adding verification of serialized data because it is small and easy to check them.
